File: backend/src/main/java/com/plugga/backend/entity/Card.java
package com.plugga.backend.entity;

import com.fasterxml.jackson.annotation.JsonIgnore;

import javax.persistence.*;
import java.sql.Timestamp;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "card")
public class Card {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "card_id")
    private int id;

    @Column(name = "last_used")
    private Timestamp lastUsed;

    @OneToMany(mappedBy = "card",
            cascade = {CascadeType.DETACH, CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH})
    private List<Side> sides;

    @OneToMany(mappedBy = "card",
            cascade = {CascadeType.DETACH, CascadeType.MERGE, CascadeType.PERSIST, CascadeType.REFRESH})
    private List<DeckCard> decks;

    public Card() {
    }

    public Card(final Timestamp lastUsed) {
        this.lastUsed = lastUsed;
    }

    public int getId() {
        return id;
    }

    public void setId(final int id) {
        this.id = id;
    }

    public Timestamp getLastUsed() {
        return lastUsed;
    }

    public void setLastUsed(final Timestamp lastUsed) {
        this.lastUsed = lastUsed;
    }

    @JsonIgnore
    public List<Side> getSides() {
        return sides;
    }

    public void setSides(final List<Side> sides) {
        this.sides = sides;
    }

    @JsonIgnore
    public List<DeckCard> getDecks() {
        return decks;
    }

    public void setDecks(final List<DeckCard> decks) {
        this.decks = decks;
    }

    @Override
    public String toString() {
        return "Card{" +
                "id=" + id +
                ", lastUsed=" + lastUsed +
                '}';
    }

    public void addSide(Side side) {
        if (sides == null) {
            sides = new ArrayList<>();
        }
        sides.add(side);
        side.setCard(this);
    }

    public void addDeck(DeckCard deck) {
        if (decks == null) {
            decks = new ArrayList<>();
        }
        decks.add(deck);
    }
}

File: backend/src/main/kotlin/com/plugga/backend/entity/Card.kt
package com.plugga.backend.entity

import com.fasterxml.jackson.annotation.JsonIdentityInfo
import com.fasterxml.jackson.annotation.JsonIdentityReference
import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.ObjectIdGenerators
import com.plugga.backend.jackson.EntityIdResolver

import javax.persistence.*
import java.sql.Timestamp

@Entity
@Table(name = "card")
@JsonIdentityInfo(
        generator = ObjectIdGenerators.PropertyGenerator::class,
        property = "id",
        resolver = EntityIdResolver::class,
        scope = Card::class
)
class Card {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "card_id")
    var id: Int = 0

    @Column(name = "last_used")
    var lastUsed: Timestamp? = null

    @OneToMany(mappedBy = "card")
    var sides: MutableList<Side>? = null

    @OneToMany(mappedBy = "card")
    @JsonIdentityReference(alwaysAsId = true)
    var decks: MutableList<DeckCard>? = null

    constructor() {}

    constructor(id: Int) {
        this.id = id
        this.lastUsed = null
        this.sides = null
        this.decks = null
    }

    override fun toString(): String {
        return "Card{" +
                "id=" + id +
                ", lastUsed=" + lastUsed +
                '}'.toString()
    }

    fun addSide(side: Side) {
        if (sides == null) {
            sides = mutableListOf()
        }
        sides!!.add(side)
        side.card = this
    }

    fun addDeck(deck: DeckCard) {
        if (decks == null) {
            decks = mutableListOf()
        }
        decks!!.add(deck)
    }
}
